code/modules/fluids/fluid_pipe_machinery.dm

```diff
@@ -0,0 +1,521 @@
+		src.UpdateIcon(TRUE)
+		return
+
+	var/datum/reagents/removed_fluid = src.pull_from_network(src.network1, src.pumprate)
+	if(!src.push_to_network(src.network2, removed_fluid))
+		removed_fluid.trans_to(network1, removed_fluid.total_volume)
+	FLICK("actuallypump", src)
+
+/obj/machinery/fluid_machinery/binary/valve
+	name = "Fluid Valve"
+	desc = "Separates two fluid pipe networks."
+	icon_state = "valve0"
+	var/on = FALSE
+
+/obj/machinery/fluid_machinery/binary/valve/attack_hand(mob/user)
+	interact_particle(user, src)
+	if(ON_COOLDOWN(src, "fluidvalve", 1 SECOND))
+		return
+	src.on = !src.on
+	src.UpdateIcon(TRUE)
+	user.visible_message(SPAN_NOTICE("[user] turns [src.on ? "on" : "off"] [src]."), SPAN_NOTICE("You turn [src.on ? "on" : "off"] [src]."))
+	if(!(src.network1 && src.network2))
+		return
+	if(src.on)
+		src.network1.merge_network(src.network2)
+	else
+		src.network1.rebuild_network()
+
+/obj/machinery/fluid_machinery/binary/valve/update_icon(animate)
+	if(animate)
+		FLICK("valve[!src.on][src.on]", src)
+	src.icon_state = "valve[src.on]"
+
+/obj/machinery/fluid_machinery/binary/valve/refresh_network(datum/flow_network/network)
+	..()
+	if(!src.on)
+		return
+	if(src.network1 && src.network2)
+		src.network1.merge_network(src.network2)
+
+/obj/machinery/fluid_machinery/binary/valve/disposing()
+	src.on = FALSE
+	src.network1.rebuild_network_force()
+	..()
+
+ABSTRACT_TYPE(/obj/machinery/fluid_machinery/trinary)
+/obj/machinery/fluid_machinery/trinary
+	var/datum/flow_network/network1
+	var/datum/flow_network/network2
+	var/datum/flow_network/network3
+
+/obj/machinery/fluid_machinery/trinary/New()
+	..()
+	switch(src.dir)
+		if(NORTH)
+			src.initialize_directions = NORTH|EAST|SOUTH
+		if(EAST)
+			src.initialize_directions = EAST|SOUTH|WEST
+		if(SOUTH)
+			src.initialize_directions = SOUTH|WEST|NORTH
+		if(WEST)
+			src.initialize_directions = WEST|NORTH|EAST
+
+/obj/machinery/fluid_machinery/trinary/disposing()
+	src.network1?.machines -= src
+	src.network2?.machines -= src
+	src.network3?.machines -= src
+	..()
+
+/obj/machinery/fluid_machinery/trinary/initialize()
+	for(var/obj/fluid_pipe/target in get_step(src, turn(src.dir, 180)))
+		if(target.initialize_directions & get_dir(target,src))
+			src.network1 = target.network
+			src.network1.machines += src
+			break
+
+	for(var/obj/fluid_pipe/target in get_step(src, turn(src.dir, -90)))
+		if(target.initialize_directions & get_dir(target,src))
+			src.network2 = target.network
+			src.network2.machines += src
+			break
+
+	for(var/obj/fluid_pipe/target in get_step(src, src.dir))
+		if(target.initialize_directions & get_dir(target,src))
+			src.network3 = target.network
+			src.network3.machines += src
+			break
+
+/obj/machinery/fluid_machinery/trinary/refresh_network(datum/flow_network/network)
+	src.network1?.machines -= src
+	src.network2?.machines -= src
+	src.network1 = null
+	src.network2 = null
+	src.initialize()
+
+/obj/machinery/fluid_machinery/trinary/filter
```

Comment:
I feel like this is either broken or I'm misunderstanding what it's supposed to do? I put a beaker of melon juice in this filter and watched it neatly filter the input melon juice into both the outlet sides(???)
<img width="294" height="178" alt="image" src="https://github.com/user-attachments/assets/a5af17aa-f306-40aa-a608-89780b18711e" />
Oil put in the same system then went straight through, so it *is* technically filtering, just in a very confusing and backwards way.